### PR TITLE
usb_dc_stm32: Minor bugfixes

### LINKED ISSUE #15443: usb_dc_stm32: Missing semaphore initialization and missing pin remapping configuration
Described in the commits of PR #9412

### PR DESCRIPTION
**Pull request by atx**
A few minor bugs I noticed when playing with an STM32F0 and the USB stack.

EDIT: Added missing S-o-b lines

Fixes #15443

### CODE REVIEW

File: drivers/usb/device/usb_dc_stm32.c [hunk 1]
```diff
@@ -351,12 +351,12 @@ int usb_dc_attach(void)
...
 	 * pair PA11/12 mapped instead of PA9/10 (e.g. stm32f070x6)
 	 */
-#if defined(CONFIG_SOC_SERIES_STM32F0X) && defined(SYSCFG_CFGR1_PA11_PA12_RMP)
+#if defined(CONFIG_USB_DC_STM32_PA11_PA12_REMAP)
 	if (LL_APB1_GRP2_IsEnabledClock(LL_APB1_GRP2_PERIPH_SYSCFG)) {
 		LL_SYSCFG_EnablePinRemap();
 	} else {
 		SYS_LOG_ERR("System Configuration Controller clock is "
-			    "disable. Unable to enable pin remapping."
+			    "disable. Unable to enable pin remapping.")
```
Comment (ydamigos) [inline drivers/usb/device/usb_dc_stm32.c]: `;` also is missing

**Review by ydamigos — CHANGES_REQUESTED** [review]
Which board did you use to test it? Could you add support for this board?

**Comment by atx** [comment]
Fixed the commits.

> Which board did you use to test it? Could you add support for this board?

I'm developing on a custom STM32F042 board. My patch for the F042 is [here](https://github.com/atx/zephyr/commit/a5c2b04ad2a6f1a9fec3fa89c5f6543894fa9f4a) but I basically just copied the F072 configs and haven't really checked for correctness.

File: drivers/usb/device/Kconfig [hunk 2]
```diff
@@ -95,4 +95,11 @@ config USB_DC_STM32_DISCONN_ENABLE
 	  Say Y if your board uses USB DISCONNECT pin to enable the
 	  pull-up resistor on USB DP.
 
+config USB_DC_STM32_PA11_PA12_REMAP
```
Comment (ydamigos) [inline drivers/usb/device/Kconfig]: I was thinking if this should be done in DT. We could add an optional property without value in the st,stm32-usb.yaml and add it to STM32F0 SoCs on QFN28 and TSSOP20 packages with USB controller. What do you think?

File: drivers/usb/device/usb_dc_stm32.c [hunk 3]
```diff
@@ -778,12 +778,21 @@ int usb_dc_ep_mps(const u8_t ep)
...
 	HAL_PCD_EP_Open(&usb_dc_stm32_state.pcd, EP0_IN, EP0_MPS, EP_TYPE_CTRL);
 	HAL_PCD_EP_Open(&usb_dc_stm32_state.pcd, EP0_OUT, EP0_MPS,
 			EP_TYPE_CTRL);
 
+	/* The DataInCallback will never be called at this point for any pending
+	 * transactions.
+	 * */
+	for (i = 0; i < CONFIG_USB_NUM_BIDIR_ENDPOINTS; i++) {
+		k_sem_init(&usb_dc_stm32_state.in_ep_state[i].write_sem, 1, 1);
+	}
```
Comment (ydamigos) [inline drivers/usb/device/usb_dc_stm32.c]: I agree that we should reset semaphores during a reset. Do we need to initialize them again? Did you try to reset them with `k_sem_reset()` instead?

The comment also is confusing. Something like "Reset IN endpoints semaphores" sounds better to me.

**Comment by atx** [comment]
I have updated the branch. Not sure if moving the pin remap to the device tree is a win for now, as it requires a fixup file (such as the [one in the olimexino_stm32](https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/olimexino_stm32/dts.fixup).

**Review by ydamigos — CHANGES_REQUESTED** [review]
>Not sure if moving the pin remap to the device tree is a win for now, as it requires a fixup file (such as the one in the olimexino_stm32).

Pin remap is related to the SoC and not the board. It is better to add something like the following in stm32f0 fixup file:
```
#ifdef ST_STM32_USB_40005C00_ENABLE_PIN_REMAP
#define CONFIG_USB_ENABLE_PIN_REMAP			ST_STM32_USB_40005C00_ENABLE_PIN_REMAP
#endif /* ST_STM32_USB_40005C00_ENABLE_PIN_REMAP */
```

**Comment by atx** [comment]
> Pin remap is related to the SoC and not the board. It is better to add something like the following in stm32f0 fixup file

Oh right, I haven't noticed that there are fixup files for individual SoCs as well.

Updated the commits.

**Review by ydamigos — CHANGES_REQUESTED** [review]
While we are waiting for @loicpoulain to check the PR, could you also fix the errors reported by shippable:
```
Commit 870f19c038:
1: UC5 Title exceeds max length (73>72): "drivers: usb_dc_stm32: Reinitialize the write semaphores on USB bus reset"
3: UC4 Line exceeds max length (79>72): "If the user attempts to send data before the USB connection is established (see"
4: UC4 Line exceeds max length (74>72): "the HID sample for an example of such code), the DataInCallback never gets"

Commit d29d76ae7c:
3: UC4 Line exceeds max length (81>72): "The SYSCFG_CFGR1_PA11_PA12_RMP define is present even on packages where the remap"
4: UC4 Line exceeds max length (75>72): "isn't strictly required. This commit makes the remap optional based on a DT"
```

**Review by ydamigos — APPROVED** [review]
@atx Please resolve the conflicts.
@loicpoulain Could you also check it?

File: drivers/usb/device/usb_dc_stm32.c (hunk 3 above)
Comment (loicpoulain) [reply drivers/usb/device/usb_dc_stm32.c]: Agree k_sem_reset seems to be the right way, avoiding unbalanced sem after reset.

**Review by loicpoulain — CHANGES_REQUESTED** [review]

**Comment by atx** [comment]
@ydamigos I have rebased on top of master
@loicpoulain Already done in the current version (altough I use k_sem_give as we need to set the semaphore to 1 on reset)

**Comment by atx** [comment]
> Sorry but I don't see any k_sem_reset prior to the k_sem_give, have you push the new commit?

Right, sorry, I misunderstood - you want to call both k_sem_reset and k_sem_give? I sort of figured that it would be redundant, as this is a binary semaphore.

Is calling k_sem_reset in an ISR safe? The [docs](https://docs.zephyrproject.org/latest/api/kernel_api.html#_CPPv311k_sem_resetP5k_sem) don't explicitly mention it.

**Comment by erwango** [comment]
@loicpoulain , can you check if @atx answer is ok for you?

**Comment by erwango** [comment]
@atx , can you rebase?

**Comment by atx** [comment]
@erwango Done

File: drivers/usb/device/usb_dc_stm32.c (hunk 3 above)
Comment (erwango) [reply drivers/usb/device/usb_dc_stm32.c]: @atx , reason not to follow @loicpoulain & @ydamigos request?

Comment (atx) [reply drivers/usb/device/usb_dc_stm32.c]: @erwango See my [earlier comment](https://github.com/zephyrproject-rtos/zephyr/pull/9412#issuecomment-424989596)

The semaphores have maximum permitted count of 1 and k_sem_give will therefore either set them to 1 or keep them at 1. Furthermore, it's unclear to me if k_sem_reset is ISR-safe as the documentation does not explicitly say so.

**Comment by loicpoulain** [comment]
ok it looks good.

**Review by loicpoulain — APPROVED** [review]

**Review by erwango — APPROVED** [review]

File: soc/arm/st_stm32/stm32f0/dts_fixup.h [hunk 4]
```diff
@@ -60,6 +60,10 @@
 #define CONFIG_USB_NUM_BIDIR_ENDPOINTS		ST_STM32_USB_40005C00_NUM_BIDIR_ENDPOINTS
 #define CONFIG_USB_RAM_SIZE			ST_STM32_USB_40005C00_RAM_SIZE
 
+#ifdef ST_STM32_USB_40005C00_ENABLE_PIN_REMAP
```
Comment (galak) [inline soc/arm/st_stm32/stm32f0/dts_fixup.h]: Why do you need the ifdef here?

**Review by galak — DISMISSED** [review]
Don't think you need the ifdef in the dts_fixup.h

File: soc/arm/st_stm32/stm32f0/dts_fixup.h (hunk 4 above)
Comment (atx) [reply soc/arm/st_stm32/stm32f0/dts_fixup.h]: #ifdefs don't do chain resolution right?

e. g.:

```
#define SOMETHING SOMETHINGELSE

#if defined(SOMETHING)
#error "Wat?"
#endif
```

Fails to build.

Comment (erwango) [reply soc/arm/st_stm32/stm32f0/dts_fixup.h]: @galak , are you good?

**Comment by aescolar** [comment]
@atx Please create a gitHub "bug" issue, labeled as medium priority, so it can be properly tracked.
I remove the "priority: medium" label from here, as PRs priorities are not looked at by anybody following our current way of working.

**Comment by carlescufi** [comment]
@atx can you please rebase this PR? Thanks!

**Comment by atx** [comment]
@aescolar  @carlescufi Rebased and created issue #15443 

Hopefully nothing is broken as I don't have time to test everything properly.

**Review by jfischer-no — APPROVED** [review]

**Pull request merged**